flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/delegation/PlannerContext.java

```diff
@@ -109,6 +109,9 @@ public PlannerContext(
 			planner.addRelTraitDef(traitDef);
 		}
 		this.cluster = FlinkRelOptClusterFactory.create(planner, new RexBuilder(typeFactory));
+
+		this.context.toRexConverterFactory_$eq(t -> new SqlExprToRexConverterImpl(
```

Comment:
Mutable member field and lazy assignement is not a good practice. I would suggest to make `FlinkContextImpl` constructor accepts a supplier function to create the SqlExprToRexConverterFactory. 

```scala
class FlinkContextImpl(
    tableConfig: TableConfig,
    functionCatalog: FunctionCatalog,
    catalogManager: CatalogManager,
    factorySupplier: Supplier[ToRexConverterFactory])
  extends FlinkContext {

  ...

  override def getToRexConverterFactory: ToRexConverterFactory = factorySupplier.get()
}
```

Then in the `PlannerContext`, we can instantiate the context early. 
```java
    public PlannerContext(...) {
       ...
       this.context = new FlinkContextImpl(
           tableConfig, 
           functionCatalog, 
           catalogManager, 
           this::createToRexConverterFactory);
       ...
    }

    private ToRexConverterFactory createToRexConverterFactory() {
		return t -> new SqlExprToRexConverterImpl(createFrameworkConfig(), typeFactory, cluster, t);
	}
```

Reply:
It is factory now, we can just remove `final` of `frameworkConfig` and `cluster`.